animations/hero-animations/index.md

```diff
@@ -0,0 +1,666 @@
+and another describing the widget in the destination route.
+From the user’s point of view, the hero appears to be shared, and
+only the programmer needs to understand this implementation detail.
+
+<aside class="alert alert-info" markdown="1">
+**Note about dialogs:**
+Heroes fly from one PageRoute to another. Dialogs
+(displayed with `showDialog()`, for example), use PopupRoutes,
+which are not PageRoutes.  At least for now,
+you can't animate a hero to a Dialog.
+For further developments (and a possible workaround), [watch this
+issue.](https://github.com/flutter/flutter/issues/10667)
+</aside>
+
+Hero animation code has the following structure:
+
+1. Define a starting Hero object, referred to as the _source
+   hero_. The hero specifies its graphical representation
+   (typically an image), and an identifying tag.
+1. Define an ending Hero object, referred to as the _destination hero_.
+   This hero also specifies its graphical representation,
+   and the same tag as the source hero.
+   It's <strong>essential that both hero objects are created with
+   the same tag</strong>, typically an object that represents the
+   underlying data. For best results, the heroes should have
+   virtually identical widget trees.
+1. Create a route that contains the destination hero.
+   The destination route defines the widget tree that exists
+   at the end of the animation.
+1. Trigger the animation by pushing the destination route on the
+   Navigator's stack. The Navigator push and pop operations trigger
+   a hero animation for each pair of heroes with matching tags in
+   the source and destination routes.
+
+Flutter calculates the tween that animates the Hero's bounds from
+the starting point to the endpoint (interpolating size and position),
+and performs the animation in an overlay.
+
+The next section describes Flutter's process in greater detail.
+
+## Behind the scenes
+
+The following describes how Flutter performs the transition from
+one route to another.
+
+<img src="images/hero-transition-0.png" alt="Before the transition the source hero appears in the source route">
+
+Before transition, the source hero waits in the source route's widget
+tree. The destination route does not yet exist, and the overlay
+is empty.
+
+---
+
+<img src="images/hero-transition-1.png" alt="The transition begins">
+
+Pushing a route to the Navigator triggers the animation.
+At t=0.0, Flutter does the following:
+
+* Lays out the destination route&mdash;off screen&mdash;using the
+  curved path as described in the Material motion spec.
+  Flutter now knows where the hero ends up.
+
+* Places the destination hero in the overlay, at the
+  same location and size as the _source_ hero.
+  Adding a hero to the overlay changes its Z-order so that
+  it appears on top of all routes.
+
+* Moves the source hero offscreen.
+
+---
+
+<img src="images/hero-transition-2.png" alt="The hero flies in the overlay to its final position and size">
```

Comment:
The diagram captions aren't quite right. The source route isn't necessarily obscured during the entirety of the hero's flight and the destination route is visible. You can see as much in your first demo.